### ISSUE TITLE
[공지] 선형 분류 실습 코드 업로드 완료

### ISSUE THREAD

**Issue by unizard** [unizard/2019.Spring.AI#34]
아래의 실습1~4문제를 혼자의 힘으로 해결해 봅시다.

오늘도 수고해 주시는 자원봉사자 조교님들께 감사의 멘트 부탁드립니다.🥳

> _본 실습자료는 모두의 딥러닝를 활용하였습니다._

<br>
<br>
<br>
<br>

## 시작하기 전 크게 한번 웃어봅시다.
[Introducing Google Tulip](https://www.youtube.com/watch?v=nsPQvZm_rgM)

**Comment by unizard** [comment]
### 실습 1

다음 데이터에 대한 **이진 분류기**를 설계하세요.
1. 아래의 데이터 x_data, y_data 로 이진 분류기를 학습하세요.
2. 학습된 모델에 학습할 때 사용했던 x_data 를 입력으로 하여, 이진 분류결과(predicted)가 y_data와 동일한지 비교하세요. 

x_data = [[1, 2],[2, 3],[3, 1],[4, 3],[5, 3],[6, 2]]
y_data = [[0],[0],[0],[1],[1],[1]]



### 힌트 
> 가설 설정

hypothesis = tf.sigmoid(tf.matmul(X, W) + b)

> 비용 함수

cost = -tf.reduce_mean(Y * tf.log(hypothesis) + (1 - Y) * tf.log(1 - hypothesis))

> 최적화 방법 설정

train = tf.train.GradientDescentOptimizer(learning_rate = 0.01).minimize(cost)

> 이진 분류 예측 결과 설정

predicted = tf.cast(hypothesis > 0.5, dtype=tf.float32)

> 이전 분류 정확도 결과 설정

accuracy = tf.reduce_mean(tf.cast(tf.equal(predicted, Y), dtype=tf.float32))

<br>
<br>

### 출력예시
Accuracy: 1.0

**Comment by unizard** [comment]
### 실습 2

당뇨병 환자를 판별하는 **이진 분류기**(AI 머신)를 설계하시오. 

- [당뇨병 환자 데이터](https://github.com/unizard/2019.Spring.AI/blob/master/diabetes.csv)를 다운받아 사용하세요. 
데이터는 당뇨병 환자를 판별하기 위한 검사 결과 8가지와 해당 결과를 기반으로 의사가 당뇨병으로 확진한 결과를 포함하고 있습니다. (Row: 환자 1명의 검사 결과, Col 1-8: 검사결과, Col 9: 당뇨병 환자 여부)
- x_data: 검사 결과, y_data: 당노병 여부를 저장하세요.
- 학습된 모델에 학습할 때 사용했던 x_data 를 입력으로 하여, 이진 분류결과(predicted)가 y_data와 동일한지 비교하세요.

### 힌트
> 가설 설정

hypothesis = tf.sigmoid(tf.matmul(X, W) + b)

> 비용 함수

cost = -tf.reduce_mean(Y * tf.log(hypothesis) + (1 - Y) * tf.log(1 - hypothesis))

> 최적화 방법 설정

train = tf.train.GradientDescentOptimizer(learning_rate=0.01).minimize(cost)

> 이진 분류 예측 결과 설정

predicted = tf.cast(hypothesis > 0.5, dtype=tf.float32)

> 이전 분류 정확도 결과 설정

accuracy = tf.reduce_mean(tf.cast(tf.equal(predicted, Y), dtype=tf.float32))

<br>
<br>

### 출력 예시
0.76679844

**Comment by Namhyunho** [comment]
늦은 시간까지 준비하시느라 고생하셨습니다.

**Comment by unizard** [comment]
### 실습 3

다음 데이터에 대한 다중 클래스 분류기를 설계하세요.

- 아래의 데이터 x_data, y_data 로 다중 클래스 분류기를 학습하세요.
- 학습된 모델에 x_data_test 를 입력으로 하여 클래스 분류 결과(y_data_test)를 확인하세요.

x_data = [[1, 2, 1, 1],[2, 1, 3, 2],[3, 1, 3, 4],[4, 1, 5, 5],[1, 7, 5, 5],[1, 2, 5, 6],[1, 6, 6, 6],[1, 7, 7, 7]]
y_data = [[0, 0, 1],[0, 0, 1],[0, 0, 1],[0, 1, 0],[0, 1, 0],[0, 1, 0],[1, 0, 0],[1, 0, 0]]

x_data_test = [[1, 11, 7, 9], [1, 3, 4, 3], [1, 1, 0, 1]]
y_data_test = 답은 무엇일까요?    
### 힌트

> 가설 설정

hypothesis = tf.nn.softmax(tf.matmul(X, W) + b)

> 비용 함수 설정

cost = tf.reduce_mean(-tf.reduce_sum(Y * tf.log(hypothesis), axis=1))

> 최적화 함수 설정

optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.1).minimize(cost)

>  테스트를 위한 입력 데이터 설정

y_data_test = sess.run(hypothesis, feed_dict={X: x_data_test})

> 출력 데이터 Hot-one-Encording 

sess.run(tf.argmax(y_data_test, 1))

<br>
<br>

### 출력예시
 [1 0 2]

**Comment by unizard** [comment]
### 실습 4

외형적인 동물의 특성 정보만을 이용하여 동물을 판별하는 **다중 클래스 분류기**를 설계하세요.

- [동물 데이터](https://github.com/unizard/2019.Spring.AI/blob/master/data-zoo-v2.csv)를 다운받아 사용하세요.
- 데이터는 동물의 종류를 판별하기 위한 특성 정보 16가지와 동물의 종류를 나타내는 클래스 번호를 담고 있습니다.
(Row: 동물 한마리의 특성 정보, Col 1-16: 특성 정보, Col 17: 동물의 종류)
- 학습된 모델에 학습할 때 사용했던 x_data 를 입력으로 하여, 다중 클래스 분류결과(predicted)가 y_data와 동일한지 비교하세요.

### 주의
- 일반 숫자로 되어 있는 클래스 라벨을 어떻게 one-hot 으로 바꿀 수 있나요?
  - 아래의 함수를 사용하세요.
  - **tf.one_hot:** https://www.tensorflow.org/api_docs/python/tf/one_hot
- one-hot 이후 텐서가 이상해요?
  - 아래의 함수를 사용하세요.
  - **tf.reshope** : https://www.tensorflow.org/api_docs/python/tf/reshape


### 힌트

> 가설 함수 설정

logits = tf.matmul(X, W) + b
hypothesis = tf.nn.softmax(logits)cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=tf.stop_gradient([Y_one_hot])))

> 비용 함수 설정

cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=tf.stop_gradient([Y_one_hot])))

> 최적화 함수 설정

optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.1).minimize(cost)

> 다중 클래스 분류 예측 결과 설정

prediction = tf.argmax(hypothesis, 1)

> 다중 클래스 분류 정확도 결과 설정

accuracy = tf.reduce_mean(tf.cast(tf.equal(prediction, tf.argmax(Y_one_hot, 1)), tf.float32))

<br>
<br>

### 출력예시
Accuracy:  100.0

**Comment by unizard** [comment]
### 실습1~4번에 대한 정답 코드
> 정답 코드 먼저 보지 마시고, 위의 실습을 혼자 힘으로 진행해 보세요.
> 혼자 힘으로 진행하실 수 있으면, 중간고사 실습 시험도 크게 어려움이 없을 겁니다.

[실습 정답 코드](https://colab.research.google.com/drive/1MWNcEaOYySXADYji1HGEUnBcwRb12g4L)

**Comment by Hongiee2** [comment]
교수님 시험공부하다 궁금한게 생겼는데요.. 실습1, 2번 모두 학습할 때 사용했던 x_data를 입력으로 이진 분류결과의 정확도를 측정했는데 1번은 1이 나오고 2번은 0.76679844이 나오는 이유를 알 수 있을까요..? 저는 2번도 1이 나올 줄 알았어요..
